src/main/java/edu/texas/social_computing/hospitals/Matching.java

```diff
@@ -88,6 +88,101 @@ public Resident getWorstAssignedResident(Hospital h) {
+               if (residentAssignment.getLocationId() != partnerAssignment.getLocationId()) {
+                   violations.add(resident.getId() + ", " + partnerId + " contains a proximity violation");
+               }
+            }
+        }
+        if (violations.size() == 0) {
+            System.out.println("Passes proximity validation");
+        } else {
+            for (String v : violations) {
+                System.out.println(v);
+            }
+        }
+    }
+
+    public void validateCapacities(List<Hospital> hospitals) {
+        List<String> violations = new ArrayList<>();
+        for (Hospital hospital : hospitals) {
+            List<Resident> assignments = getAssignedResidents(hospital);
+            Integer assignmentSize = assignments.size();
+            Integer capacity = hospital.getCapacity();
+            if (assignmentSize > capacity) {
+                violations.add(hospital.getId() + " exceeds capacity: " + assignmentSize + " > " + capacity);
+            }
+        }
+        if (violations.size() == 0) {
+            System.out.println("Passes capacity validation");
+        } else {
+            for (String v : violations) {
+                System.out.println(v);
+            }
+        }
+    }
+
+    private boolean isRankedHigherThanWorstMatch(Hospital hospital, String residentId) {
+        Resident worstMatch = getWorstAssignedResident(hospital);
+        List<String> hosPrefs = hospital.getPreferences();
+        if (hosPrefs.indexOf(residentId) == -1) {
+            return false;
+        }
+        if (worstMatch == null) {
+            return true;
+        }
+        String worstResId = worstMatch.getId();
+        int hosResPrefIndex = hosPrefs.indexOf(residentId);
+        int hosWorstPrefIndex = hosPrefs.indexOf(worstResId);
+        if (hosResPrefIndex < hosWorstPrefIndex) {
+            return true;
+        }
+        return false;
+    }
+
+    public void validateStability(List<Resident> residents, HospitalTable hospitalTable) {
+        List<String> violations = new ArrayList<>();
+        for (Resident resident : residents) {
+            String resId = resident.getId();
+            List<String> resPrefs = resident.getPreferences();
+            Hospital residentAssignment;
+            if (hasAssignment(resident)) {
+                residentAssignment = getAssignedHospital(resident);
+                int resPrefIndex = resPrefs.indexOf(residentAssignment.getId());
+                if (resPrefIndex > 0) {
+                    for (int i = 0; i < resPrefIndex; i++) {
+                        Hospital h = hospitalTable.getHospitalById(resPrefs.get(i));
+                        if (isRankedHigherThanWorstMatch(h, resId)) {
+                            violations.add("(" + resId + ", " + h.getId() + ") is a blocking pair" );
+                        }
+                    }
+                }
+            } else {
+                for (int i = 0; i < resPrefs.size(); i++) {
```

Comment:
```
for (String pref : resPrefs) {
                    Hospital h = hospitalTable.getHospitalById(pref);
                    if (isRankedHigherThanWorstMatch(h, resId)) {
                        violations.add("(" + resId + ", " + h.getId() + ") is a blocking pair" );
                    }
```
i'd reformat as above for readability.
I'd probably add a break in here because just 1 violation for a person is enough and having it return more than one violation for each resident may not be very helpful. If it is helpful maybe we should figure out how to determine which one is the 'worst' violation.

Reply:
this one can be ignored if you take my comment for the if-else block.